sinks/grpsink/state_helper_test19.go

```diff
@@ -0,0 +1,63 @@
+// +build go1.9
+
+package grpsink
+
+import (
+	"context"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+
+	"google.golang.org/grpc/connectivity"
+)
+
+func waitThroughStateSequence(t *testing.T, sink *GRPCStreamingSpanSink, dur time.Duration, seq ...connectivity.State) {
+	t.Helper()
+
+	first, current := seq[0], sink.grpcConn.GetState()
+	if !assert.ObjectsAreEqual(first, current) {
+		t.Fatalf("Wanted %q for initial connection state, but got %q", first, current)
+	}
+
+	for i, state := range seq[:len(seq)] {
+		ctx, cf := context.WithTimeout(context.Background(), dur)
+		if !sink.grpcConn.WaitForStateChange(ctx, state) {
+			t.Fatalf("(seq %v) Connection never transitioned from %q", i, state.String())
+		}
+		cf()
+	}
+}
+
+func waitThroughFiniteStateSequence(t *testing.T, sink *GRPCStreamingSpanSink, dur time.Duration, seq ...connectivity.State) {
+	t.Helper()
+
+	waitThroughStateSequence(t, sink, dur, seq[:len(seq)-1]...)
+
+	last, current := seq[len(seq)-1], sink.grpcConn.GetState()
+	if !assert.ObjectsAreEqual(last, current) {
+		t.Fatalf("Wanted %q for final connection state, but got %q", last, current)
+	}
+}
+
+func reconnectWithin(t *testing.T, sink *GRPCStreamingSpanSink, dur time.Duration) {
+	t.Helper()
+	ctx, cf := context.WithTimeout(context.Background(), dur)
+	for {
+		state := sink.grpcConn.GetState()
+		switch state {
+		case connectivity.Ready:
+			// Spin on the internal state marker that indicates the stream state is bad
+			for !atomic.CompareAndSwapUint32(&sink.bad, 0, 0) {
```

Comment:
shouldn't this check that the context timeout has been exceeded too? (Otherwise it'll spin forever/until the test timeout)

Reply:
(now done, but github didn't notice)